compiler/compiler.go

```diff
@@ -0,0 +1,213 @@
+	"strings"
+
+	"github.com/alecthomas/protobuf/parser"
+	pb "google.golang.org/protobuf/types/descriptorpb"
+)
+
+type fdBuilder struct {
+	proto3   bool
+	fileDesc *pb.FileDescriptorProto
+}
+
+type messageBuilder struct {
+	proto3      bool
+	fullName    string
+	messageDesc *pb.DescriptorProto
+}
+
+func NewFileDescriptor(file string, p *parser.Proto) *pb.FileDescriptorProto {
+	b := &fdBuilder{
+		proto3: p.Syntax == "proto3",
+	}
+	b.buildFileDescriptor(file, p)
+	return b.fileDesc
+}
+
+func (b *fdBuilder) buildFileDescriptor(file string, p *parser.Proto) {
+	b.fileDesc = newFileDescritpor(file, b.proto3)
+	for _, e := range p.Entries {
+		b.addEntry(e)
+	}
+}
+
+func newFileDescritpor(file string, proto3 bool) *pb.FileDescriptorProto {
+	var syntax *string
+	if proto3 {
+		s := "proto3"
+		syntax = &s
+	}
+	return &pb.FileDescriptorProto{
+		Name:   &file,
+		Syntax: syntax,
+	}
+}
+
+func (b *fdBuilder) addEntry(e *parser.Entry) {
+	switch {
+	case e.Package != "":
+		b.fileDesc.Package = &e.Package
+	case e.Import != nil:
+		b.buildImports(e.Import)
+	case e.Message != nil:
+		m := newMessage(e.Message, b.proto3, "")
+		b.fileDesc.MessageType = append(b.fileDesc.MessageType, m)
+	case e.Enum != nil:
+	case e.Service != nil:
+	case e.Option != nil:
+	case e.Extend != nil:
+	default:
+		panic("cannot interpret Entry")
+	}
+}
+
+func (b *fdBuilder) buildImports(imp *parser.Import) {
+	b.fileDesc.Dependency = append(b.fileDesc.Dependency, imp.Name)
+	if imp.Public {
+		idx := int32(len(b.fileDesc.Dependency) - 1)
+		b.fileDesc.PublicDependency = append(b.fileDesc.PublicDependency, idx)
+	}
+}
+
+func newMessage(m *parser.Message, proto3 bool, outerTypes string) *pb.DescriptorProto {
+	fullName := m.Name
+	if outerTypes != "" {
+		fullName = outerTypes + "." + m.Name
+	}
+	b := &messageBuilder{
+		proto3:   proto3,
+		fullName: fullName,
+	}
+	b.buildMessageDescriptor(m)
+	return b.messageDesc
+}
+
+func (b *messageBuilder) buildMessageDescriptor(pm *parser.Message) {
+	b.messageDesc = &pb.DescriptorProto{Name: &pm.Name}
+	for _, e := range pm.Entries {
+		b.addEntry(e)
+	}
+}
+
+func (b *messageBuilder) addEntry(e *parser.MessageEntry) {
+	switch {
+	case e.Field != nil:
+		b.buildField(e.Field)
+	case e.Message != nil:
+		m := newMessage(e.Message, b.proto3, b.fullName)
+		b.messageDesc.NestedType = append(b.messageDesc.NestedType, m)
+	case e.Enum != nil:
+	case e.Option != nil:
+	case e.Oneof != nil:
+	case e.Extend != nil:
+	case e.Reserved != nil:
+	case e.Extensions != nil:
+	default:
+		panic("cannot interpret MessageEntry")
+	}
+}
+
+func (b *messageBuilder) buildField(pField *parser.Field) {
+	fieldDesc := b.createField(pField)
+	b.messageDesc.Field = append(b.messageDesc.Field, fieldDesc)
+
+	if b.proto3 && fieldDesc.Proto3Optional != nil && *fieldDesc.Proto3Optional {
+		idx := int32(len(b.messageDesc.Field)) - 1
+		fieldDesc.OneofIndex = &idx
+		name := "_" + *fieldDesc.Name
+		dcl := &pb.OneofDescriptorProto{Name: &name}
```

Comment:
Huh is that how proto3 optionals are encoded?

Reply:
It seems like it 🤷‍♀️ - I'm just trying to align protoc Filedescriptors with this.